DM-44095: Implement queryDimensionRecords for RemoteButler

dhirving opened the pull request:
Implemented a large subset of `registry.queryDimensionsRecords` for `RemoteButler` by forwarding to the new query system.

## Checklist

- [ ] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Improve type safety of ParserYacc usage

ParserYacc is untyped, so wrap its usage in a type-safe function to isolate the "type: ignore" to a single file.

dhirving committed:
Partially implement expression strings for QueryTree

Add an initial implementation of expression string parsing for the new query system.  This does not yet include full support for time/timespan comparisons.

dhirving committed:
Implement queryDimensionRecords for RemoteButler

dhirving committed:
Add more unit tests for queryDimensionRecords

This revealed some dataset-type-handling discrepancies between the original implementation and the new one, so fix that too.

dhirving commented on python/lsst/daf/butler/queries/_expression_strings.py:
```diff
@@ -0,0 +1,296 @@
+        tree = parse_expression(expression)
+    except Exception as exc:
+        raise InvalidQueryError(f"Failed to parse expression '{expression}'") from exc
+
```
The original implementation does some additional processing on the tree here.  I'm not sure if that is stuff already handled by the new query system's validation logic or if we need to handle it here too.

```
    expr = NormalFormExpression.fromTree(tree, NormalForm.DISJUNCTIVE)
    # Check the expression for consistency and completeness.
    visitor = CheckVisitor(data_id, dimensions, bind, defaults, allow_orphans=allow_orphans)
```

TallJimbo replied:
This stuff should indeed already be handled.

dhirving commented on python/lsst/daf/butler/queries/_expression_strings.py:
```diff
@@ -0,0 +1,296 @@
+_VisitorResult: TypeAlias = Predicate | _ColExpr | _Null | _RangeLiteral | _Sequence
+
+
+class _ConversionVisitor(TreeVisitor[_VisitorResult]):
```
The vast majority of the coverage for the query expression language is on `queryDataIds`, so I'm going to implement that next instead of trying to replicate the tests using `queryDimensionRecords`.

The main unimplemented features here are `IN` and miscellaneous `Time`/`Timespan` conversions in `visitBinaryOp`.

There is missing test coverage in the original implementation as well, so I will backfill the tests and complete this implementation in a follow-up to the implementation of queryDataIds.

TallJimbo replied:
Did `IN` and the `Time`/`Timespan` operators look like they were impossible to represent in the new `Predicate` tree, or did you just not want to do them yet?

Another way to approach the testing is that the coverage for `Predicate` trees and their mapping to SQL should now be pretty decent via the new query system's tests, so just testing the conversion from string to `Predicate` tree is probably sufficient for poking at all of the edge cases.

dhirving replied:
At first glance they appear to be relatively straightforward to implement -- there are just a lot of cases and no test coverage, so I wanted to defer to a later PR with less going on.

dhirving commented on python/lsst/daf/butler/registry/tests/_registry.py:
```diff
@@ -3212,18 +3236,19 @@ def do_query(element, datasets=None, collections=None):
             Test("visit", "-visit.name", (2, 1)),
-            Test("visit", "day_obs,-timespan.begin", (2, 1)),
-        )
+            Test("visit", "day_obs,-visit.timespan.begin", (2, 1)),
```
When day_obs became its own dimension, this query became ambiguous because `timespan` is present in both day_obs and visit (with different meaning/values).  The new query system correctly threw an exception complaining about the ambiguity, but the old one silently picks the visit one.

dhirving commented on python/lsst/daf/butler/registry/tests/_registry.py:
```diff
@@ -3386,7 +3423,7 @@ def testExposureQueries(self):
                 record.id
                 for record in registry.queryDimensionRecords("visit", instrument="HSC")
-                .order_by("id")
+                .order_by("visit")
```
Similar to above, `id` became ambiguous when `day_obs` became a dimension.

dhirving commented on python/lsst/daf/butler/registry/tests/_registry.py:
```diff
@@ -106,6 +107,22 @@ class RegistryTests(ABC):
     """True if the registry class being tested supports regex searches for
     collections."""
 
+    supportsDetailedQueryExplain: bool = True
```
This is mostly stuff that requires `LIMIT 1` execution of subqueries to diagnose.  This is marked as a TODO in DirectQueryDriver, so we can flip this feature flag to get these tests when someone goes to backfill that feature.

dhirving commented on python/lsst/daf/butler/registry/tests/_registry.py:
```diff
@@ -106,6 +107,22 @@ class RegistryTests(ABC):
+    to query methods.
+    """
+
+    supportsQueryGovernorValidation: bool = True
```
I think this just wasn't implemented for the new query system?  Not sure how important/desirable this validation is -- I think it's more consistent to just get no results here instead of sometimes throwing an exception depending which field you tried to filter on.

TallJimbo replied:
We've gone back and forth on whether to raise or get no results here, due to conflicting requests from users (which we didn't always recognize as conflicting).  For the new query system I came down on the side of just returning no results, both because I also think that's more consistent and because in the new system (especially thinking about `RemoteButler`) it looked harder to have the exception raised at the point at which an error message would be most useful.

I've added this to the list of changes to be RFC'd on [DM-43587](https://rubinobs.atlassian.net/browse/DM-43587).

TallJimbo commented on python/lsst/daf/butler/queries/_expression_strings.py:
```diff
@@ -0,0 +1,296 @@
+            case _:
+                raise AssertionError("Unhandled expression constant")
+
+        element, column = categorizeElementId(self.universe, name)
```
I'd prefer to delegate to `lsst.daf.butler.queries.convert_args.interpret_identifier` here.  That will require initializing the visitor with more information about the state of the query the expression is being added to, but that'll solve the problem of how to handle `INGEST_DATE`, provide hopefully-better error messages, and make `where` and `order_by` identifier intepretation more consistent.

(I actually think it will only recognize lower-case `ingest_date`, so we may need some case-wrangling in `interpret_identifier).

TallJimbo commented on python/lsst/daf/butler/queries/_expression_strings.py:
```diff
@@ -0,0 +1,296 @@
+        return expression
+
+    def visitPointNode(self, ra: _VisitorResult, dec: _VisitorResult, node: Node) -> _VisitorResult:
+        raise NotImplementedError("POINT() function is not supported yet")
```
I'm pretty confident it's never actually been supported, at least not beyond the parser.

TallJimbo commented on python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py:
```diff
@@ -0,0 +1,137 @@
+            # We have to eagerly fetch the results to prevent
+            # leaking the resources associated with QueryDriver.
+            records = list(result)
+        yield from records
```
I suspect (but do not know) that `return iter(records)` is a little more efficient in cases like this.

TallJimbo commented on python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py:
```diff
@@ -0,0 +1,137 @@
+            if a.dataId:
+                query = query.where(a.dataId)
+            if a.kwargs:
+                query = query.where(a.kwargs)
```
We should combine these last two cases into one:
```
if a.dataId:
     kwargs = a.kwargs if a.kwargs else {}  # or use default_factory=dict in dataclass
     query = query.where(a.dataId, **a.kwargs)
```
This lets the values in `kwargs` replace those in `dataId`, which is what users have come to expect from data IDs in kwargs, instead of putting the conflicting constraint into the query and dooming it yield no results.

dhirving committed:
Move interpret_identifier to its own file

This function is needed in convert_expression_string_to_predicate, so it needs to be in another file to avoid a circular import.

dhirving committed:
Use interpret_identifier in where string parsing

Remove some duplicated code by deferring to interpret_identifier when parsing identifiers in expression strings.  This function gives better diagnostics, and it can use the query context to handle things like the 'INGEST_DATE' constant.

dhirving committed:
Fix kwargs in RemoteButler queryDimensionRecords

kwargs passed to queryDimensionRecords are supposed to override dataID values, instead of both constraints being applied separately.
